addons/purchase/tests/test_purchase_downpayment.py

```diff
@@ -0,0 +1,686 @@
+        dp_invoice = self.purchase_order.invoice_ids[0]
+        dp_invoice.invoice_date = datetime.datetime.today()
+        dp_invoice.action_post()
+
+        self.assertRecordValues(pol_downpayment, [{
+            'price_unit': 587.5,
+            'discount': 0.0,
+            'product_qty': 0.0,
+            'qty_invoiced': 1.0,
+            'qty_to_invoice': -1.0,
+        }])
+        self.assertEqual(self.purchase_order.invoice_status, 'to invoice')
+
+        self.make_downpayment(advance_payment_method='delivered')
+
+        po_invoice = max(self.purchase_order.invoice_ids)
+        self.assertEqual(len(po_invoice.invoice_line_ids.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))),
+                         len(self.purchase_order.order_line.filtered(lambda l: not (l.display_type == 'line_section' and l.name == "Down Payments"))), 'All lines should be invoiced')
+        self.assertEqual(len(po_invoice.invoice_line_ids.filtered(lambda l: l.display_type == 'line_section' and l.name == "Down Payments")), 1, 'A single section for downpayments should be present')
+        self.assertEqual(po_invoice.amount_total, self.purchase_order.amount_total - pol_downpayment.price_unit, 'Downpayment should be applied')
+        po_invoice.invoice_date = datetime.datetime.today()
+        po_invoice.action_post()
+
+        credit_note_wizard = self.env['account.move.reversal'].with_context({'active_ids': [po_invoice.id], 'active_id': po_invoice.id, 'active_model': 'account.move'}).create({
+            'reason': 'reason test refund with downpayment',
+            'journal_id': po_invoice.journal_id.id,
+        })
+        credit_note_wizard.refund_moves()
+        invoice_refund = self.purchase_order.invoice_ids.sorted(key=lambda inv: inv.id, reverse=False)[-1]
+        invoice_refund.invoice_date = datetime.datetime.today()
+        invoice_refund.action_post()
+
+        self.assertEqual(self.pol_product_order.qty_to_invoice, 5.0, "As the refund still exists, the quantity to invoice is the ordered quantity")
+        self.assertEqual(self.pol_product_order.qty_invoiced, 0.0, "The qty invoiced should be zero as, with the refund, the quantities cancel each other")
+        self.assertEqual(len(self.pol_product_order.invoice_lines), 2, "The product line is invoiced, so it should be linked to 2 invoice lines (invoice and refund)")
+        self.assertEqual(pol_downpayment.qty_to_invoice, -1.0, "As the downpayment was invoiced separately, it will still have to be deducted from the total invoice (hence -1.0), after the refund.")
+        self.assertEqual(pol_downpayment.qty_invoiced, 1.0, "The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is")
+        self.assertEqual(len(pol_downpayment.invoice_lines), 3, "The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)")
```

Comment:
if I understand this test correctly, these should be:
`"As the refund still exists, the quantity to invoice is the ordered quantity"` => `"The refund should make it so the quantity to invoice is the ordered quantity"`

`"The qty invoiced should be zero as, with the refund, the quantities cancel each other"` => `"The qty invoiced should be zero since the refund cancels the previously invoiced amount"`

`"As the downpayment was invoiced separately, it will still have to be deducted from the total invoice (hence -1.0), after the refund."` => "Downpayment was invoiced separately should still count as invoiced after refund of the product line. Since the ordered qty is 0 for down payments, this means -1 is remaining to invoice."

I would move the assert for the `qty_invoiced` for the dp lines above the `qty_to_invoice` assert since it makes the checks  easier to understand in a logical order
`""The qty to invoice should be 1 as, with the refund, the products are not invoiced anymore, but the downpayment still is""` => `"The qty invoiced should remain 1 since the refund was only for the ordered product, not the downpayment"`

"The product line is invoiced, so it should be linked to 3 invoice lines (downpayment invoice, partial invoice and refund)" => I'm actually confused now why the downpayment is linked to the partial invoice and refund. Not sure if I reworded the ones above it correctly and I'm not sure how this one should be reworded since I don't understand if product line = downpayment line or not in this assert